### ISSUE TITLE
add `eslint-import-resolver-typescript` as a dependency

### ISSUE THREAD

**Issue by Zamiell** [un-ts/eslint-plugin-import-x#150]
From @silverwind:

> I think eslint-import-resolver-typescript is currently only mandatory for typescript, but I think we should make it a dependency and default to it for both JS and TS. The default JS resolver has too many issues like not supporting exports object format which are resolved in the typescript resolver.

(Originally in [this pull request](https://github.com/un-ts/eslint-plugin-import-x/pull/122).)

Doing this would simplify the config for the end-user, which would be a big win!

cc @SukkaW @michaelfaith

**Comment by SukkaW** [comment]
There are still a few serious issues with `eslint-import-resolver-typescript`.

Also I am putting my bet on https://github.com/9romise/eslint-import-resolver-oxc as well.

**Comment by michaelfaith** [comment]
Thougts about absorbing typescript `parser` into the typescript flat config, too?  That'd add an optional peer dependency on `@typescript-eslint/parser`, but remove the need for users to have to define a parser.

**Comment by SukkaW** [comment]
> That'd add an optional peer dependency

I don't know how the package managers would handle this. It seems that `yarn 1`, `yarn@berry`, `npm@5`, `npm@latest`, `pnpm@6`, and `pnpm@latest` all handle optional peer dependencies differently.

Also, since most package managers don't hoist the transitive dependencies.

**Comment by michaelfaith** [comment]
> > That'd add an optional peer dependency
> 
> I don't know how the package managers would handle this. It seems that `yarn 1`, `yarn@berry`, `npm@5`, `npm@latest`, `pnpm@6`, and `pnpm@latest` all handle optional peer dependencies differently.
> 
> Also, since most package managers don't hoist the transitive dependencies.

Not sure I follow.  All three package managers support `peerDependenciesMeta.optional` (yarn 1 & berry, npm since v7 / node 15, and pnpm).  It wouldn't be a transitive dependency though.  The consuming project would still need to have it installed in as a first class dep, if they're using the typescript config.

**Comment by SukkaW** [comment]
> It wouldn't be a transitive dependency though.

What happens if someone is publishing his eslint presets to npm (e.g. `@antfu/eslint-config` and `eslint-config-sukka`)? Then `eslint-import-resolver-typescript` becomes a transitive dependency of that preset package. Even if `eslint-import-resolver-typescript` is listed as a direct dependency of that preset package, it is still a transitive dependency (`current project -> eslint-config-presets => eslint-import-resolver-typescript`).

This used to cause me shit loads of issues when I am building `eslint-config-sukka`.

**Comment by Zamiell** [comment]
> > It wouldn't be a transitive dependency though.
> 
> What happens if someone is publishing his eslint presets to npm (e.g. `@antfu/eslint-config` and `eslint-config-sukka`)? Then `eslint-import-resolver-typescript` becomes a transitive dependency of that preset package. Even if `eslint-import-resolver-typescript` is listed as a direct dependency of that preset package, it is still a transitive dependency (`current project -> eslint-config-presets => eslint-import-resolver-typescript`).
> 
> This used to cause me shit loads of issues when I am building `eslint-config-sukka`.

i tried this over the weekend, and with the npm package manager specifically, it won't work properly if the dependency chain is like:

my-project --> my-linting-metapackage --> eslint-config-foo --> eslint-plugin-import-x & eslint-import-resolver-typescript

However, the dependency chain DOES work just fine if the dependency chain is like:

my-project --> my-linting-metapackage --> eslint-config-foo & eslint-import-resolver-typescript

I didn't test any other package managers than npm.

Thus, I think there is still value in putting `eslint-import-resolver-typescript` as a direct dependency of `eslint-plugin-import-x`, since some people (or maybe most people? unsure) will not consume `eslint-import-plugin-x` through a linting metapackage that nests the `eslint-import-resolver-typescript` dep more than two layers deep, if that makes sense. (And you could add this limitation to the readme.)

**Comment by silverwind** [comment]
> I was wondering if it would make sense to add support for (conditional) exports fields in package json files to the (default) node resolver instead of introducing another dependency?

It would, but the counter-argument is that maintaining will not go well in the long run and it's better to focus efforts on making one resolver good for everthing, even at the cost of additional dependencies, which are often not much of a cost to bear.

**Comment by error-four-o-four** [comment]
Please let me know if this is unrelated or not the right place (as there also is #40) or if I'm wrong.
eslint-import-resolver-node (and others) rely on [resolve](https://www.npmjs.com/package/resolve) whereas eslint-import-resolver-typescript introduces [enhanced-resolve](https://www.npmjs.com/package/enhanced-resolve). In my case (and I don't know if it's an edge case) I'm pretty sure that I won't be using webpack in any fo my projects the near future. I didn't take a closer look at these packages regarding performance or benchmark but I'm sure that their doing a great job.
Anyway your point is absolutely reasonable.
The way I see it is, that it would be great to have some kind of extendable resolver. Like a monorepo with a 'core' package. Other resolver packages/workspaces (like ts, webpack, vue, whatever) could extend this core package and wouldn't have to rewrite/build core functionality again.

**Comment by vladshcherbin** [comment]
Something should be done with the default resolver, which is not aware of `exports` definition in `package.json` https://github.com/import-js/eslint-plugin-import/issues/1810 - more and more packages give incorrect error warnings e.g. [eslint-plugin-perfectionist](https://www.npmjs.com/package/eslint-plugin-perfectionist), [typescript-eslint](https://www.npmjs.com/package/typescript-eslint)

To say at least, I'd absolute love seeing a huge warning notice about this limitation and discouragement of using default resolver, would save so much time and head scratching

<img width="755" alt="image" src="https://github.com/user-attachments/assets/0f30bb8c-ca26-4ffb-8275-19c471850623">

**Comment by oleksandr-danylchenko** [comment]
Hello, @SukkaW 👋🏻 
Any news on when could we expect the version where the resolver with the `exports` support will become a default?

**Comment by SukkaW** [comment]
> Hello, [@SukkaW](https://github.com/SukkaW?rgh-link-date=2025-03-04T11%3A17%3A36.000Z) 👋🏻 Any news on when could we expect the version where the resolver with the `exports` support will become a default?

In the next major version, the built-in node resolver (that supports `exports`) will become the default resolver.

In the meantime, you will still need to manually import and use that in your `eslint.config.js`.

**Comment by SukkaW** [comment]
> > > Why not use `import.meta.resolve` for the node resolver? Is it because it's synchronous only?
> > 
> > 
> > Yes. The resolving happens inside the ESLint plugin, and the ESLint plugin needs to be synchronous. Besides `import.meta.resolve` only exists in ESM while both `eslint-plugin-import-x` and `eslint-plugin-import` have many users working with CJS.
> 
> [@SukkaW](https://github.com/SukkaW?rgh-link-date=2025-03-06T02%3A24%3A10.000Z) See also https://www.npmjs.com/package/@dual-bundle/import-meta-resolve

`import.meta.resolve` doesn't work well with CJS-only modules since the `import-meta-resolve` package strictly implements the ESM resolving algorithm, while `import.meta.resolve` follows the ESM resolving algorithm which misses a few crucial steps for CJS resolving.

Dropping CJS-only modules support would be a breaking change we might not want to introduce.

**Comment by JounQin** [comment]
> `import.meta.resolve` doesn't work well with CJS-only modules since the `import-meta-resolve` package strictly implements the ESM resolving algorithm, while the ESM resolving algorithm misses a few crucial steps for CJS resolving.

Can you elaborate what's the issue here for resolving CJS-only modules?

```ts
import {resolve} from 'import-meta-resolve'

// A file:
console.log(resolve('./index.js', import.meta.url))
//=> file:///Users/tilde/Projects/oss/import-meta-resolve/index.js

// A CJS package:
console.log(resolve('builtins', import.meta.url))
//=> file:///Users/tilde/Projects/oss/import-meta-resolve/node_modules/builtins/index.js

// A scoped CJS package:
console.log(resolve('@eslint/eslintrc', import.meta.url))
//=> file:///Users/tilde/Projects/oss/import-meta-resolve/node_modules/@eslint/eslintrc/lib/index.js

// A package with an export map:
console.log(resolve('micromark/lib/parse', import.meta.url))
//=> file:///Users/tilde/Projects/oss/import-meta-resolve/node_modules/micromark/lib/parse.js

// A node builtin:
console.log(resolve('fs', import.meta.url))
//=> node:fs
```

**Comment by SukkaW** [comment]
> > `import.meta.resolve` doesn't work well with CJS-only modules since the `import-meta-resolve` package strictly implements the ESM resolving algorithm, while the ESM resolving algorithm misses a few crucial steps for CJS resolving.
> 
> [@SukkaW](https://github.com/SukkaW?rgh-link-date=2025-03-06T09%3A10%3A13.000Z) Can you elaborate what's the issue here for resolving CJS-only modules?

Check out the test case I introduced in https://github.com/un-ts/eslint-plugin-import-x/pull/209/files#diff-8f1ca7dc1831cb622e7f670ebf6b8bf8fe0ce43bd506ae9a11c93b90491ea6d5. Actually I have already tried `@dual-bundle/import-meta-resolve` when implementing #209, turns out that it can't pass the test cases, so I switch to `enhanced-resolve` instead.

**Comment by JounQin** [comment]
> Check out the test case I introduced in https://github.com/un-ts/eslint-plugin-import-x/pull/209/files#diff-8f1ca7dc1831cb622e7f670ebf6b8bf8fe0ce43bd506ae9a11c93b90491ea6d5. Actually I have already tried `@dual-bundle/import-meta-resolve` when implementing #209, turns out that it can't pass the test cases, so I switch to `enhanced-resolve` instead.

Thanks, I'll take a look when I'm free.

**Comment by JounQin** [comment]
By the way, `eslint-import-resolver-typescript` is now also powered by [rspack-resolver](https://github.com/unrs/rspack-resolver) which has P'n'P support correctly.

So actually you don't need `eslint-import-resolver-oxc` any longer.

**Comment by JounQin** [comment]
Close in favor of #272

**Comment by Zamiell** [comment]
For future readers of this thread, note that this issue was not resolved. If you use TypeScript, you still need to install the `eslint-import-resolver-typescript` package in addition to the `eslint-plugin-import-x` package, or else you will get warnings from ESLint like this:

```
Warning:   1:1  warning  Resolve error: typescript with invalid interface loaded as resolver  import-x/export
```

@JounQin Sorry for the ping, but just to clarify: was this closed as not planned? Or should we link to some other new issue?

**Comment by JounQin** [comment]
@Zamiell Yeah, the default resolver is always `node`, but after thinking, we'd have a chance to mark `eslint-import-resolver-typescript` as an optional peer dependency for the `typescript` config, we should add a special case to exclude some known packages as resolver.

By the way, `eslint-import-resolver-typescript` also marks `eslint-plugin-import-x` as optional peer dependency, I don't whether it would be a problem.

**Comment by silverwind** [comment]
Why not change the default resolver in a breaking change? As far as I'm aware `eslint-import-resolver-typescript` works universally for everything, including node code and js.

**Comment by JounQin** [comment]
@silverwind `eslint-import-resolver-typescript` preferrs `.ts`/`.d.ts`, etc over the default `node` resolver, and also automatic `tsconfig.json`/`jsconfig.json` support.

I did indeed want to change it as a `typescript: true` option instead at #40.

**Comment by silverwind** [comment]
> eslint-import-resolver-typescript preferrs .ts/.d.ts, etc over the default node

But is it an issue for a JS-only user? I think not. Also node since v22 supports executing typescript via `--experimental-strip-types` etc, so node code is no longer JS-only.

**Comment by JounQin** [comment]
> But is it an issue for a JS-only user? I think not. Also node since v22 supports executing typescript via `--experimental-strip-types` etc, so node code is no longer JS-only.

1. Node 20 is still in LTS, and we target Node 18 for now.
2. AFAIK, `--experimental-strip-types` only works for non node_modules files, while we're parsing `.d.ts` definitions.

**Comment by silverwind** [comment]
Yes, it's still a limitation of `--experimental-strip-types` that it does not work for files inside a `node_modules` folder. I do hope that they lift this restriction soon.

**Comment by JounQin** [comment]
By the way, I think the native TypeScript support doesn't affect the default Node resolver at all, even for `--experimental-strip-types`, `.ts` extensions are still required, and that already works today.

**Comment by silverwind** [comment]
Yes indeed, file extensions always were and are still required in Node's resolver. They are optional in typescript only.

**Comment by lishaduck** [comment]
> By the way, `eslint-import-resolver-typescript` also marks `eslint-plugin-import-x` as optional peer dependency, I don't whether it would be a problem.

It's not a problem for most pms other than that it's very weird to do, but I know deno doesn't handle it correctly: https://github.com/denoland/deno/issues/27544.

> By the way, I think the native TypeScript support doesn't affect the default Node resolver at all, even for `--experimental-strip-types`, `.ts` extensions are still required, and that already works today.

So for `erasableSyntaxOnly`/`rewriteRelativeImportSpecifiers`, you don't need `eslint-import-resolver-typescript`, or do you still need it for `.d.ts` resolution (for `type` imports)?

**Comment by JounQin** [comment]
> or do you still need it for `.d.ts` resolution (for `type` imports)?

Good point, it worth thinking how to resolve it, I'm not quite sure now. We'll need to check whether it's a `type` import/export to determine `.d.ts` or `.js` should be preferred.

**Comment by lishaduck** [comment]
> > or do you still need it for `.d.ts` resolution (for `type` imports)?
> 
> Good point, it worth thinking how to resolve it, I'm not quite sure now. We'll need to check whether it's a `type` import/export to determine `.d.ts` or `.js` should be preferred.

And, of course, without `verbatimModuleSyntax` that's a whole can of worms...

**Comment by JounQin** [comment]
> And, of course, without `verbatimModuleSyntax` that's a whole can of worms...

I believe it's not supported by Node itself.

**Comment by lishaduck** [comment]
> > And, of course, without `verbatimModuleSyntax` that's a whole can of worms...
> 
> I believe it's not supported by Node itself.

Ah, makes sense 👍
I only use `verbatimModuleSyntax` so I haven't actually had to mess with that.
